armi/reactor/converters/axialExpansionChanger/assemblyAxialLinkage.py

```diff
@@ -57,118 +73,187 @@ def _determineLinked(componentA, componentB):
+        # Use islice to avoid making intermediate lists of subsequences of blocks
+        lower = itertools.chain((None,), itertools.islice(blocks, 0, nBlocks - 1))
+        upper = itertools.chain(itertools.islice(blocks, 1, None), (None,))
+        links = {}
+        for low, mid, high in zip(lower, blocks, upper):
+            links[mid] = AxialLink(lower=low, upper=high)
+        return links
+
     def _determineAxialLinkage(self):
         """Gets the block and component based linkage."""
         for b in self.a:
-            self._getLinkedBlocks(b)
-            for c in getSolidComponents(b):
+            for c in iterSolidComponents(b):
                 self._getLinkedComponents(b, c)
 
-    def _getLinkedBlocks(self, b):
-        """Retrieve the axial linkage for block b.
-
-        Parameters
-        ----------
-        b : :py:class:`Block <armi.reactor.blocks.Block>`
-            block to determine axial linkage for
-
-        Notes
-        -----
-        - block linkage is determined by matching ztop/zbottom (see below)
-        - block linkage is stored in self.linkedBlocks[b]
-         _ _
-        |   |
-        | 2 |  Block 2 is linked to block 1.
-        |_ _|
-        |   |
-        | 1 |  Block 1 is linked to both block 0 and 1.
-        |_ _|
-        |   |
-        | 0 |  Block 0 is linked to block 1.
-        |_ _|
-        """
-        lowerLinkedBlock = None
-        upperLinkedBlock = None
-        block_list = self.a.getChildren()
-        for otherBlk in block_list:
-            if b.name != otherBlk.name:
-                if b.p.zbottom == otherBlk.p.ztop:
-                    lowerLinkedBlock = otherBlk
-                elif b.p.ztop == otherBlk.p.zbottom:
-                    upperLinkedBlock = otherBlk
-
-        self.linkedBlocks[b] = [lowerLinkedBlock, upperLinkedBlock]
-
-        if lowerLinkedBlock is None:
-            runLog.debug(
-                "Assembly {0:22s} at location {1:22s}, Block {2:22s}"
-                "is not linked to a block below!".format(
-                    str(self.a.getName()),
-                    str(self.a.getLocation()),
-                    str(b.p.flags),
-                ),
-                single=True,
-            )
-        if upperLinkedBlock is None:
-            runLog.debug(
-                "Assembly {0:22s} at location {1:22s}, Block {2:22s}"
-                "is not linked to a block above!".format(
-                    str(self.a.getName()),
-                    str(self.a.getLocation()),
-                    str(b.p.flags),
-                ),
-                single=True,
-            )
+    def _findComponentLinkedTo(
+        self, c: Component, otherBlock: typing.Optional[Block]
+    ) -> typing.Optional[Component]:
+        if otherBlock is None:
+            return None
+        candidate = None
+        # Iterate over all solid components in the other block that are linked to this one
+        areLinked = functools.partial(self.areAxiallyLinked, c)
```

Comment:
I trust this works, but I don't know how it does 😅

Reply:
[`functools.partial`](https://docs.python.org/3/library/functools.html#functools.partial) is a way to do a lot of things, but I usually use it to make a function that I'm going to call a lot with similar arguments. This could have been replaced in the `filter` call with a lambda expression
```python
for otherComp in filter(lambda other: self.areAxiallyLinked(c, other), iterSolidComponents(otherBlock)):
...
```
The benefit (or at least my perceived benefit) is we don't need to "re-lookup" via `.`-access `self.areAxiallyLinked` at each step through the iteration.